# Introduction to Python programming.

### [Gerard Gorman](http://www.imperial.ac.uk/people/g.gorman), [Christian Jacobs](http://christianjacobs.uk/)

### Updated for MPECDT by [David Ham](http://www.imperial.ac.uk/people/david.ham)

# Lecture 1: Computing with formulas

## Learning objectives:

* Execute a Python statement from within IPython.
* Learn what a program variable is and how to express a mathematical expression in code.
* Print program outputs.
* Access mathematical functions from a Python module.

# Programming a mathematical formula

Here is a formula for the position of a ball in vertical motion, starting at ground level (i.e. $y=0$) at time $t=0$:
     $$ y(t) = v_0t- \frac{1}{2}gt^2 $$

where:

* $y$ is the height (position) as a function of time $t$
* $v_0$ is the initial velocity (at $t=0$)
* $g$ is the acceleration due to gravity

The computational task is: given $v_0$, $g$ and $t$, compute the value $y$. 

**How do we program this task?** A program is a sequence of instructions given to the computer. However, while a programming language is much **simpler** than a natural language, it is more **pedantic**. Programs must have correct syntax, i.e., correct use of the computer language grammar rules, and no misprints.

So let's execute a Python statement based on this example. Evaluate $y(t) = v_0t- \frac{1}{2}gt^2$ for $v_0=5$, $g=9.81$ and $t=0.6$. If you were doing this on paper you would probably write something like this: $$ y = 5\cdot 0.6 - {1\over2}\cdot 9.81 \cdot 0.6^2.$$ Happily, writing this in Python is very similar:

print(5*0.6 - 0.5*9.81*0.6**2)

Go ahead and mess with the code above to see what happens when you change values and rerun. To see what I mean about programming being pedantic, see what happens if you replace `**` with `^`:

In [3]:
print(5*0.6 - 0.5*9.81*0.6**2)

1.2342


or `write` rather than `print`:

In [4]:
write 5*0.6 - 0.5*9.81*0.6**2

SyntaxError: invalid syntax (<ipython-input-4-b7024092bb96>, line 1)

While a human might still understand these statements, they do not mean anything to the Python interpreter. Rather than throwing your hands up in the air whenever you get an error message like the above (you are going to see many during the course of these lectures!!!) train yourself to read the message patiently to get an idea what it is complaining about and re-read your code from the perspective of the pedantic Python interpreter.

Error messages can look bewildering (frustrating etc.) at first, but it gets much **easier with practise**.

## Storing numbers in variables
From mathematics you are already familiar with variables (e.g. $v_0=5,\quad g=9.81,\quad t=0.6,\quad y = v_0t -{1\over2}gt^2$) and you already know how important they are for working out complicated problems. Similarly, you can use variables in a program to make it easier to read and understand.

In [2]:
v0 = 5
g = 9.81
t = 0.6
y = v0*t - 0.5*g*t**2
print(y)

1.2342


In [1]:
a=2
print(type(a))
a=2.5
print(type(a))

<class 'int'>
<class 'float'>


This program spans several lines of text and uses variables, otherwise the program performs the same calculations and gives the same output as the previous program.

In mathematics we usually use one letter for a variable, resorting to using the Greek alphabet and other characters for more clarity. The main reason for this is to avoid becoming exhausted from writing when working out long expressions or derivations. However, when programming you should use more descriptive names for variable names. This might not seem like an important consideration for the trivial example here but it becomes increasingly important as the program gets more complicated and if someone else has to read your code. **Good variable names make a program easier to understand!**

Permitted variable names include:

* One-letter symbols.
* Words or abbreviation of words.
* Variable names can contain a-z, A-Z, underscore ("'_'") and digits 0-9, **but** the name cannot start with a digit.
* In Python 3, variable names can also include letters from other alphabets, such as α or π.

Variable names are case-sensitive (i.e. "'a'" is different from "'A'"). Let's rewrite the previous example using more descriptive variable names:

In [5]:
initial_velocity = 5
g = 9.81
TIME = 0.6
VerticalPositionOfBall = initial_velocity*TIME - 0.5*g*TIME**2
print(VerticalPositionOfBall)
from math import pi
print(pi)

1.2342
3.141592653589793


In [3]:
from math import pi as π
radius = 2
area = π * radius ** 2
print(area)

12.566370614359172


Certain words have are **reserved** in Python and **cannot be used as variable names**. These are: *and, as, assert, break, class, continue, def, del, elif, else, except, exec, finally, for, from, global, if, import, in, is, lambda, not, or, pass, print, raise, return, try, with, while,* and *yield*.

## Adding comments to code

Not everything written in a computer program is intended for execution. In Python anything on a line after the '#' character is ignored and is known as a **comment**. You can write whatever you want in a comment. Comments are intended to be used to explain what a snippet of code is intended for. It might for example explain the objective or provide a reference to the data or algorithm used. This is both useful for you when you have to understand your code at some later stage, and indeed for whoever has to read and understand your code later.

In [9]:
# Program for computing the height of a ball in vertical motion.
v0 = 5    # Set initial velocity in m/s.
g = 9.81  # Set acceleration due to gravity in m/s^2.
t = 0.6   # Time at which we want to know the height of the ball in seconds.
y = v0*t - 0.5*g*t**2 # Calculate the vertical position
print(y)

1.2342


## <span style="color:blue">Exercise: Convert from meters to Imperial length units</span>
Make a program where you set a length given in meters and then compute and write out the corresponding length measured in inches, in feet, in yards, and in miles. Use the fact that one inch is 2.54 cm, one foot is 12 inches, one yard is 3 feet, and one British mile is 1760 yards. As a verification, a length of 640 meters corresponds to 25196.85 inches, 2099.74 feet, 699.91 yards, or 0.3977 miles.

In [16]:
a=2
cin=0.0254
cfo=12*cin
cyar=3*cfo
cmil=1760*cyar
ain=a/cin
print(ain)
afo=a/cfo
print(afo)
ayar=a/cyar
print(ayar)
amil=a/cmil
print(amil)

78.74015748031496
6.561679790026248
2.187226596675416
0.001242742384474668


## Formatting numbers as strings
Often we want to print out results using a combination of text and numbers, e.g. "'At t=0.6 s, y is 1.23 m'". Particularly when printing out floating point numbers we should **never** quote numbers to a higher accuracy than they were measured. Python provides a *printf formatting* syntax exactly for this purpose. We can see in the following example that the *slot* `%g` was used to express the floating point number with the minimum number of significant figures, and the *slot* `%.2f` specified that only two digits are printed out after the decimal point.

In [15]:
print("At t=%gs, y is %.2fm." % (t, y))
print(t)
print("""il mio trisnonno quando nacque 
aveva t=%0.2f anni""" % t)

At t=0.6s, y is 1.23m.
0.6
il mio trisnonno quando nacque 
aveva t=0.60 anni


Notice in this example how the values in the tuple `(t, y)` are inserted into the *slots*.

Sometimes we want a multi-line output. This is achieved using a triple quotation (*i.e.* `"""`):

In [18]:
print("""At t=%f s, a ball with
initial velocity v0=%.3E m/s
is located at the height %.2f m.
""" % (t, v0, y))

At t=0.600000 s, a ball with
initial velocity v0=5.000E+00 m/s
is located at the height 1.23 m.



## <span style="color:blue">Exercise: Compute the air resistance on a football</span>
The drag force, due to air resistance, on an object can be expressed as
$$F_d = \frac{1}{2}C_D\rho AV^2$$
where $\rho$ is the density of the air, $V$ is the velocity of the object, $A$ is the cross-sectional area (normal to the velocity direction), and $C_D$ is the drag coefficient, which depends heavily on the shape of the object and the roughness of the surface.</br></br>
The gravity force on an object with mass $m$ is $F_g = mg$, where $g = 9.81ms^{−2}$.</br></br>
Write a program that computes the drag force and the gravity force on an object. Write out the forces with one decimal in units of Newton ($N = kgm/s^2$). Also print the ratio of the drag force and the gravity force. Define $C_D$, $\rho$, $A$, $V$, $m$, $g$, $F_d$, and $F_g$ as variables, and put a comment with the corresponding unit.</br></br>
As a computational example, you can initialize all variables with values relevant for a football kick. The density of air is $\rho = 1.2 kg m^{−3}$. For any ball, we have obviously that $A = \pi a^2$, where $a$ is the radius of the ball, which can be taken as $11cm$ for a football. The mass of the ball is $0.43kg$. $C_D$ can be taken as $0.2$.</br></br>
Use the program to calculate the forces on the ball for a hard kick, $V = 120km/h$ and for a soft kick, $V = 10km/h$ (it is easy to make the mistake of mixing inconsistent units, so make sure you compute with V expressed in m/s). Make sure you use the *printf* formatting style introduced above.

In [33]:
from math import pi 
rho=1.2 #density of air
a=0.11 #radius of a ball in m
m=0.43 #mass of the ball in kg
cd=0.2 #drag coefficient
v_hard= 120/3.6 #velocity from hard kick in m/s
v_soft= 10/3.6 #velocity from soft kick in m/s
A=pi*a**2
g=9.81 #gravity acceleration
Fg=m*g
print("""The gravity force is Fg=%0.1f N""" % (Fg))
Fd_hard=0.5*cd*rho*A*v_hard**2
print("""The drag force from a hard kick is Fdhard=%0.1f N""" % (Fd_hard))
Fdsoft=0.5*cd*rho*A*vsoft**2
print("""The drag force from a soft kick is Fdsoft=%0.4f N""" % (Fdsoft))

The gravity force is Fg=4.2 N
The drag force from a hard kick is Fdhard=5.1 N
The drag force from a soft kick is Fdsoft=0.0352 N


## How are arithmetic expressions evaluated?
Consider the random mathematical expression, ${5\over9} + 2a^4/2$, implemented in Python as `5.0/9 + 2*a**4/2`.

The rules for evaluating the expression are the same as in mathematics: proceed term by term (additions/subtractions) from the left, compute powers first, then multiplication and division. Therefore in this example the order of evaluation will be:

1. `r1 = 5.0/9`
2. `r2 = a**4`
3. `r3 = 2*r2`
4. `r4 = r3/2`
5. `r5 = r1 + r4`

Use parenthesis to override these default rules. Indeed, many programmers use parenthesis for greater clarity.

## <span style="color:blue">Exercise: Compute the growth of money in a bank</span>
Let *p* be a bank's interest rate in percent per year. An initial amount *A* has then grown to $$A\left(1+\frac{p}{100}\right)^n$$ after *n* years. Write a program for computing how much money 1000 euros have grown to after three years with a 5% interest rate.

In [17]:
A = 1000 # initial savings in euros
p = 0.05 # interest rate per year
n = 3    # number of years
G = A*(1+p)**n-A
print(G)


157.62500000000023


## Standard mathematical functions
What if we need to compute $\sin x$, $\cos x$, $\ln x$, etc. in a program? Such functions are available in Python's *math module*. In fact there is a vast universe of functionality for Python available in modules. We just *import* in whatever we need for the task at hand.

In this example we compute $\sqrt{2}$ using the *sqrt* function in the *math* module:

In [36]:
import math
r = math.sqrt(2)
print(r)

1.4142135623730951


or:

In [35]:
from math import sqrt
r = sqrt(2)
print(r)

1.4142135623730951


or:

In [39]:
from math import *   # import everything in math
r = sqrt(2)
print(r)
z=tanh(3)
print(z)


1.4142135623730951
0.9950547536867305


Another example:

In [37]:
from math import sin, cos, log
x = 1.2
print(sin(x)*cos(x) + 4*log(x))   # log is ln (base e)

1.0670178174513938


## <span style="color:blue">Exercise: Evaluate a Gaussian function</span>
The bell-shaped Gaussian function,
$$f(x)=\frac{1}{\sqrt{2\pi}s}\exp\left(-\frac{1}{2} \left(\frac{x-m}{s}\right)^2\right)$$
is one of the most widely used functions in science and technology. The parameters $m$ and $s$ are real numbers, where $s$ must be greater than zero. Write a program for evaluating this function when $m = 0$, $s = 2$, and $x = 1$. Verify the program's result by comparing with hand calculations on a calculator.

In [41]:
dir(math)
help(math.acos)

Help on built-in function acos in module math:

acos(...)
    acos(x)
    
    Return the arc cosine (measured in radians) of x.

